django/http/response.py

```diff
@@ -588,8 +588,9 @@ class JsonResponse(HttpResponse):
     An HTTP response class that consumes data to be serialized to JSON.
 
     :param data: Data to be dumped into json. By default only ``dict`` objects
-      are allowed to be passed due to a security flaw before ECMAScript 5. See
-      the ``safe`` parameter for more information.
+      are allowed to be passed due to a security flaw before ECMAScript 5.
+      Also, dict objects make the expansion of the response simpler. See the
+      ``safe`` parameter for more information.
```

Comment:
IMO, it's not clear to me what the second-last sentence means. And then when you go to the `safe` parameter section for more info, it doesn't expand on that, it just repeats the same sentence.

Is it trying to say that even though only `dict` objects are allowed, it isn't a bad thing?

Reply:
There is a description about how to disable the default behaviour (`By default only ``dict`` objects are allowed to be passed`) in the `safe` parameter section.

> Is it trying to say that even though only dict objects are allowed, it isn't a bad thing?

Yes, I think so

Reply:
No, you should only return dict objects in json responses unless you're reimplementing an existing protocol that forces a JSON Array on you

Reply:
This was historically because of a security issue, but now it's because arrays do not support forward comparability

Reply:
After reconsideration, I would leave `JsonResponse`'s docstring and the warning in _"Serializing non-dictionary objects"_ untouched. Instead we can add a small note in the main text, e.g.
```diff
diff --git a/docs/ref/request-response.txt b/docs/ref/request-response.txt
index a34880a4be..4e09e3ec5c 100644
--- a/docs/ref/request-response.txt
+++ b/docs/ref/request-response.txt
@@ -1092,6 +1092,10 @@ parameter to ``False``::
 
 Without passing ``safe=False``, a :exc:`TypeError` will be raised.
 
+Note that an API based on ``dict`` objects is more extensible, flexible, and
+easier to maintain forwards compatibility. Therefore, you should avoid using
+non-dict objects in JSON-encoded response.
+
 .. warning::
 
     Before the `5th edition of ECMAScript
```